wit/batch.wit

```diff
@@ -1,27 +1,81 @@
 /// A keyvalue interface that provides batch operations.
+/// 
+/// A batch operation is an operation that operates on multiple keys at once.
+/// 
+/// Batch operations are useful for reducing network round-trip time. For example, 
+/// if you want to get the values associated with 100 keys, you can either do 100 get
+/// operations or you can do 1 batch get operation. The batch operation is
+/// faster because it only needs to make 1 network call instead of 100.
+/// 
+/// A batch operation does not guarantee atomicity, meaning that if the batch
+/// operation fails, some of the keys may have been modified and some may not. 
+/// Transactional operations are being worked on and will be added in the future to
+/// provide atomicity.
+/// 
+/// Data consistency in a key value store refers to the gaurantee that once a 
+/// write operation completes, all subsequent read operations will return the
+/// value that was written.
+/// 
+/// The level of consistency in batch operations can vary depending on the 
+/// implementation. This interface does not guarantee strong consistency, meaning
+/// that if a write operation completes, subsequent read operations may not return
+/// the value that was written.
```

Comment:
Given the above wording that frames `batch` operations as-if they were batches of multiple `readwrite` operations, I would expect the consistency of `batch ` operations to be defined in terms of a batch of `readwrite` operations (without ordering between the operations in the batch, I assume).  The `readwrite` interface doesn't seem to talk about the consistency of operations, so that might be a useful complementary addition.  FWIW, there's a whole [spectrum of consistency choices](https://jepsen.io/consistency), with production implementations offering [multiple options](https://learn.microsoft.com/en-us/azure/cosmos-db/consistency-levels), so it's probably worth thinking a lot about which ones we want and maybe finding some experts in this domain.

Reply:
My impression of consistency model for `readwrite` is that we may want [Eventual Consistency](https://en.wikipedia.org/wiki/Eventual_consistency) but the best we can do is left this to the implementor of keyvalue stores to determine what consistency and isolation options they provide.

Reply:
I think if we leave it up to the implementor, then the client of the interface (the component developer) has to assume the worst possible case (b/c if the code assumes stronger consistency than the implementor implements, their app breaks).  Now it's certainly fine to specify very weak consistency (like I did for `wasi:keyvalue/cache`, except adding durability).  However, developers may sometimes need stronger consistency (that's why, e.g., CosmosDB offers stronger consistency), in which case we need to define a **separately-named interface** (which may or may not have the same type signature) which is well-defined to offer that stronger consistency.  That way, a component can declare (as part of its static interface) exactly what the component is assuming (and even import multiple keyvalue stores with different consistency, which can be useful when you have different requirements for different data).  This approach is in contrast to leaving this consistency implicit and configured on a host-by-host basis, which I believe would certainly lead to subtly-broken components.  But it does mean we need to think carefully about how many and which `interface`s to provide.

Reply:
I agree to your point that if the consistency level is implicit and configured on the data stores / host, then components that assume stronger consistency level may break. I have a few points to your proposal for **separately-named interface** for varying consistency levels. 

Consistency as a term in data systems is a pretty overloaded term. It could mean "data consistency" in asynchronously replicated systems, or the "C" in ACID that refers to relational databases that are in good state. In discussing consistency in distributed key-value stores, we primarily refer to data consistency in asynchronously replicated systems. Here, the consistency model determines the validity of concurrent read/write operations. For example, an eventual consistent system guarantees that if we pause the updates to the system, the system eventually will return the last updated value for read. A strongly consistent system, conversely, gives the impression of a singular data copy, always returning the most current value.

Given this definition, it's difficult for me to see a well-defined mapping between the APIs (CRUD operations) of a keyvalue store and the underlying consistency model it assumes. That is, Whether the data store is strongly consistent or eventual consistent doesn not really change the API that much. If there is no significant changes to the APIs, I would prefer to leave `consistency-model` as a parameter to the operations in `readwrite` interface. 

To prove my point, I checked the APIs of DynamoDB and CosmosDB, two of the major public cloud offering: [DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html) has a parameter called `ConsistentRead` for read operations. If it's true, then the consistency model is strongly consistent. [CosmosDB](https://learn.microsoft.com/en-us/rest/api/cosmos-db/get-a-document)'s `getDocument` operation has a header parameter for setting `consistency-model` to one of the four: "Strong, Bounded, Session, or Eventual". 

My second point is that the necessity of a parameter like`consistency-model` in `readwrite` interface is debatable to begin with. The reason that I say this is

For one, there isn't a well-defined list of consistent models out there respected by the majority of the keyvalue stores except, maybe, eventual consistency. CosmosDB might be the only mainstream kv store that offers such a fine-grained consistent model options. DynamoDB only offers two: eventual and strong. [`etcd`](https://github.com/etcd-io/etcd/blob/v3.3.25/etcdserver/etcdserverpb/rpc.proto#L404) offers "serializable" and "linearizable". Apache Zookeeper ensures strong conssitency because of its define, so it does not allow for different levels to be set for individual read operations.

Extending on the first point, if we do allow for fine-grained consistency levels that component can target to, we will suffer from limited portability among data stores. The consistency model itself, when lives at the abstraction layer, has a clear definition of what it means to be "strongly consistent", but the devils are in the details of the implementation by different data stores. What they claim to be strongly consistent may come with conditions. For example, DynamoDB claims **strong consitent read** but it only applies to single-key operation, which means if two keys are updated concurrently, a strongly consistent read might not reflect these updates in the same order across all nodes. A component that targets strong consistency model has limited runtime / host to executes, and even worse, it still might break and introduce surprises. 

Thid point, some data stores allow for fine-grained control at the layer of `readwrite` interface, while others only provide settings at a more global level (out of the scope for wasi:keyvalue). For example, Redis Cluster provides some level of consistency through replication and sharding but this is more about the architecture than individual operation-level settings. On the other side, I can see `Cassandra`, `DynamoDB`, `MongoDB` and `Couchbase` provide mechanisms to control consistency at the op level. In the case of global mechanism, I am not entirely sure what to map to with an operation that sets to a certain consistency level.

Reply:
Thanks for the thoughtful reply; you make a lot of good points.

> Given this definition, it's difficult for me to see a well-defined mapping between the APIs (CRUD operations) of a keyvalue store and the underlying consistency model it assumes.

In some sense, the consistency model is a part of the semantic definition of the APIs (CRUD operations), because the consistency model says, when you perform some sequence of operations, what may or must happen.

> That is, Whether the data store is strongly consistent or eventual consistent doesn not really change the API that much.

You sortof answer this question at the end of your second comment: yes, mostly the operations would be the same, but there may be certain operations that only make sense with a certain consistency model.  But if we do see a high degree of duplication, we could think of adding an `include` to `interface` (allowing one `interface` to include the contents of another `interface`, symmetric to what you added to `world`) to reduce that.

> If there is no significant changes to the APIs, I would prefer to leave consistency-model as a parameter to the operations in readwrite interface.

I think you make the counterpoint to this below: since we're trying to define interfaces that work over many distinct kv-store implementations and since some of the kv stores don't allow specifying consistency dynamically as part of the operation, we cover more implementations by having consistency specified statically as part of the `interface`.  Also, knowing the required consistency statically can end up being really useful when determining where the component can run (near the writer vs. near any (geo-distributed) replica). 

> A component that targets strong consistency model has limited runtime / host to executes, and even worse, it still might break and introduce surprises.

If code needs (as in, can't work without) a strongly-consistent keyvalue store, there's really no substitute for us specifying an interface with strong consistency and the host then providing an implementation that actually implements strong consistency.  There can always be bugs, of course, but with a clear declaration as part of the interface, it's at least possible to say who's at fault and who needs to fix their bug.

That all being said, it's always an option for the wasi-keyvalue proposal to start with *only* eventual consistency (and therefore say that client code must *never* depend on getting stronger consistency).  It's just important that we say so clearly to (at least attempt to) set developer expectations appropriately.

Reply:
Based on this discussion, it sounds like the suggestion is to have an atomic batch interface as well as softer consistency batch interface?

Reply:
Oh sorry, I took this thread on a bit of a tangent.  My root comment was suggesting that, since this comment defines the consistency of `batch` in terms of multiple `readwrite` operations, it seems like we should define the consistency of `readwrite` (which then opened the sub-question of: should we have multiple `readwrite` interfaces that capture different consistency models, since there's not one "right" answer).  But in any case, no, I wasn't suggesting we have an atomic batch interface (this is probably not efficiently implementable in many systems) -- I think what you have now for `batch` makes sense.

Reply:
Great I will incorporate the comment into this pr.

Reply:
I've added eventual consistency guarantee to the readwrite interface. I will open an issue for your question "should we have multiple readwrite interfaces that capture different consistency models".